server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/VariableWidthHistogramAggregator.java

```diff
@@ -0,0 +1,378 @@
+                for (int i = 0; i < clusterCentroids.size() - 1; i++) {
+                    avgBucketDistance += clusterCentroids.get(i + 1) - clusterCentroids.get(i);
+                }
+                avgBucketDistance /= clusterCentroids.size() - 1;
+            }
+        } else {
+            int bucketOrd = getNearestBucket(val);
+            double distance = Math.abs(clusterCentroids.get(bucketOrd)- val);
+            if(bucketOrd == -1 || distance > (2 * avgBucketDistance) && clusterCentroids.size() < shardSize) {
+                // Make a new bucket since the document is distant from all existing buckets
+                // TODO: Create a new bucket even if (# buckets) > shardSize. This will require merging nearby buckets
+                //       after sorting by centroid in buildAggregation(), to ensure at most shardSize buckets are produced
+                collectBucket(sub, doc, clusterCentroids.size());
+                addToCluster((int)clusterCentroids.size(), val);
+            } else {
+                collectBucket(sub, doc, bucketOrd);
+                addToCluster(bucketOrd, val);
+            }
+        }
+    }
+
+    /**
+     * Add <code>val</code> to the cluster represented by the bucket at index <code>bucketOrd</code>.
+     * This requires recalculating the cluster's centroid, min, max, and size.
+     */
+    private void addToCluster(int bucketOrd, double val){
+        assert bucketOrd <= clusterCentroids.size(); // We don't want to add ord n + 1 if ord n doesn't exist yet
+
+        final BigArrays bigArrays = context.bigArrays();
+
+        double bucketMin;
+        double bucketMax;
+        if(clusterCentroids.size() == bucketOrd){
+            // The bucket doesn't exist yet -> Allocate space for it
+            clusterMaxes = bigArrays.resize(clusterMaxes, bucketOrd + 1);
+            clusterMins = bigArrays.resize(clusterMins, bucketOrd + 1);
+            clusterCentroids = bigArrays.resize(clusterCentroids, bucketOrd + 1);
+            clusterSizes = bigArrays.resize(clusterSizes, bucketOrd + 1);
+
+            bucketMax = val;
+            bucketMin = val;
+            clusterSizes.set(bucketOrd, 0);
+        } else {
+            bucketMax = Math.max(clusterMaxes.get(bucketOrd), val);
+            bucketMin = Math.min(clusterMins.get(bucketOrd), val);
+        }
+
+        // Recalculate the centroid
+        double oldCentroid = clusterCentroids.get(bucketOrd);
+        double clusterSize = clusterSizes.get(bucketOrd);
+        double newCentroid = ((oldCentroid * clusterSize) + val) / (clusterSize + 1);
+
+        clusterMaxes.set(bucketOrd, bucketMax);
+        clusterMins.set(bucketOrd, bucketMin);
+        clusterCentroids.set(bucketOrd, newCentroid);
+        clusterSizes.increment(bucketOrd, 1);
+    }
+
+    private void bucketCachedDocs() throws IOException{
+        Integer[] indexes = new Integer[(int)cachedValues.size()];
```

Comment:
Could this be `int[]`?

Also I wonder if it is worth bumping the breaker for these. If they are more than a couple of kb it *probably* is. If you made them with `BigArrays` closed them when you were done then it'd all happen automatically. But I don't believe we have a `sort` on `BigIntArray` so I'm not sure what the best thing to do here is.

Reply:
Below I sort `indexes` by the corresponding cached values, so I use a comparator. I don't think you can use sort with comparators on primitives, so I'm not sure how this can be done without using `Integer[]`. 

`BigIntArray` would be nice, but we need to sort so I don't think we can use it.

Reply:
Yeah, I see that. So, Lucene has a whole bunch of classes that extend from `Sorter` that let you apply a sorting algorithm to arbitrary stuff. `InPlaceMergeSorter` might be a good base for this because it is simple. They have a `TimSorter` but it is much more complex. Or you could skip it and leave a TODO and we can get to it later. It isn't a big deal, but it is a bit squicky to me to do all the `Integer` "fun" just to get a comparator.